spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -95,7 +96,12 @@ case class RewriteDelete(spark: SparkSession) extends Rule[LogicalPlan] with Rew
       case _ =>
         // apply hash partitioning by file if the distribution mode is hash or range
         val numShufflePartitions = conf.numShufflePartitions
-        RepartitionByExpression(Seq(fileNameCol), remainingRowsPlan, numShufflePartitions)
+        val rbeCtor = DistributionAndOrderingUtils.repartitionByExpressionCtor
+        if (Spark3VersionUtil.isSpark30) {
+          rbeCtor.newInstance(Seq(fileNameCol), remainingRowsPlan, Integer.valueOf(numShufflePartitions))
+        } else {
+          rbeCtor.newInstance(Seq(fileNameCol), remainingRowsPlan, Some(numShufflePartitions))
+        }
```

Comment:
I think that we should have a set of util methods for creating nodes that change across versions rather than embedding this logic everywhere. That way, we know that all of our calls to create a node like `RepartitionByExpression` go through the same path that is compatible. That would also keep the reflection code in one place rather than in many different classes.

Could you create a util class for those methods? Something like `CatalystPlans.newRepartitionByExpression`.

Reply:
Currently, there are 3 catalyst classes used in Iceberg whose constructors changed signatures between Spark 3.0 and 3.1:
`o.a.s.sql.catalyst.expressions.Alias`
`o.a.s.sql.catalyst.expressions.SortOrder`
`o.a.s.sql.catalyst.plans.logical.RepartitionByExpression`

The first 2 have `createAlias` and `createSortOrder` methods respectively (in `o.a.s.sql.catalyst.utils.RewriteRowLevelOperationHelper` and `o.a.s.sql.catalyst.utils.DistributionAndOrderingUtils` respectively) for constructing them.
I added `createRepartitionByExpression` in `o.a.s.sql.catalyst.utils.PlanUtils`.

Since there is an existing util object for catalyst plans, I put the method there. I'm fine with naming it `newRepartitionByExpression` but I used `createX` since I was using that naming scheme before (for the expressions). I'm also open to moving the methods for the expressions into one util object if you prefer, but that could also be done later.

Reply:
Good choices. Let's move the others later (after the release) and I support using the same conventions as existing methods.